File: 2050-count-good-numbers/count-good-numbers.ts
function countGoodNumbers(n: number): number {
    const mod: bigint = 1000000007n;

    // use fast exponentiation to calculate x^y % mod
    function quickmul(x: bigint, y: bigint): bigint {
        let ret: bigint = 1n;
        let mul: bigint = x;
        while (y > 0n) {
            if (y % 2n === 1n) {
                ret = (ret * mul) % mod;
            }
            mul = (mul * mul) % mod;
            y = y / 2n;
        }
        return ret;
    }

    return Number(
        (quickmul(5n, BigInt(n + 1) / 2n) * quickmul(4n, BigInt(n) / 2n)) % mod,
    );
}

File: 2050-count-good-numbers/count-good-numbers.java
class Solution {
    static final long MOD = 1000000007;

    public int countGoodNumbers(long n) {
        long evenPositions = (n + 1) / 2; // ceil(n/2)
        long oddPositions = n / 2;        // floor(n/2)

        long part1 = modPow(5, evenPositions, MOD);
        long part2 = modPow(4, oddPositions, MOD);

        return (int)((part1 * part2) % MOD);
    }

    // Fast modular exponentiation
    private long modPow(long base, long exp, long mod) {
        long result = 1;
        base %= mod;

        while (exp > 0) {
            if ((exp & 1) == 1) { // if exp is odd
                result = (result * base) % mod;
            }
            base = (base * base) % mod;
            exp >>= 1; // divide exp by 2
        }

        return result;
    }
}
